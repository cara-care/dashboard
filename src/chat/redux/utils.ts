import { capitalize } from 'lodash';
import { ChatMessage, ChatUser } from '.';

export const findLastContact = (
  messages: ChatMessage[],
  username: string | undefined
) => {
  const lastContact = messages.find((msg) => msg.author !== username);
  return capitalize(lastContact?.author);
};

export const deletedUserData = (id: number, username: string): ChatUser => ({
  id,
  username,
  age: -1,
  allergies: [],
  dateJoined: '',
  enrolledProgrammes: [],
  diagnosis: '',
  lastSeen: '',
  platform: '',
  sex: '',
  email: '',
  timezone: '',
  nickname: '',
});
