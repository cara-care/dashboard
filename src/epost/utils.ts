const timeZoneOptions = { timeZone: 'Europe/Berlin' };

export const germanFormattedDate = (draftDate?: string) => {
  if (!draftDate) {
    return new Date().toLocaleString('de-DE', timeZoneOptions);
  }
  return new Date(draftDate).toLocaleString('de-DE', timeZoneOptions);
};
