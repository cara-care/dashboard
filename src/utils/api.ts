import axios from 'axios';

// In production the requests made to `/api/` are proxied to production backend -> `public/_redirects`:
// https://www.netlify.com/docs/redirects/#proxying
// In development the reuqests are proxied to staging backend in package.json:
// https://create-react-app.dev/docs/proxying-api-requests-in-development#docsNav
const api = axios.create({
  baseURL: process.env.NODE_ENV === 'production' ? '/api/' : '/',
  timeout: 5500,
  headers: {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  },
});

function getCookie(name: string) {
  let cookieValue = null;
  if (document.cookie && document.cookie !== '') {
    const cookies = document.cookie.split(';');
    for (let i = 0; i < cookies.length; i++) {
      const cookie = cookies[i].trim();
      // Does this cookie string begin with the name we want?
      if (cookie.substring(0, name.length + 1) === name + '=') {
        cookieValue = decodeURIComponent(cookie.substring(name.length + 1));
        break;
      }
    }
  }
  return cookieValue;
}

api.interceptors.request.use(
  // Set CSRF token before sending a request
  function (config) {
    const csrfToken = getCookie('csrftoken');
    if (csrfToken) {
      config.headers.common['X-CSRFToken'] = csrfToken;
    } else {
      config.headers.common['X-CSRFToken'] = null;
    }
    return config;
  },
  (error) => Promise.reject(error)
);

export const login = ({
  username,
  password,
}: {
  username: string;
  password: string;
}) => api.post<{}>('/dashboard/login/', { username, password });

export const logout = () => {
  return api.delete('/dashboard/logout/');
};

export const getMe = () => {
  return api.get('/dashboard/me/');
};

export const getUserByEmail = (email: string) =>
  api.post<{ id: number; nickname: string; timezone: string }>(
    '/dashboard/find-user/',
    { email }
  );

export const getTrackingDataPoints = ({
  userId,
  start,
  end,
  limit,
  offset,
}: {
  userId?: number | null;
  start: string;
  end: string;
  limit: number;
  offset: number;
}) => {
  return api.get<
    {
      id: number;
      name: string;
      started: Date;
      completed: Date;
    }[]
  >(
    `/dashboard/${
      userId || 'me'
    }/data-points/?start=${start}&end=${end}&limit=${limit}&offset=${offset}`
  );
};

export const downloadMealItemImage = ({
  userId,
  realmId,
}: {
  userId?: number | null;
  realmId: string;
}) => {
  return api.get(`/dashboard/${userId || 'me'}/images/${realmId}/`, {
    responseType: 'blob',
  });
};

export const getQuestionnaires = ({
  userId,
  limit,
  offset,
}: {
  userId: number;
  limit: number;
  offset: number;
}) => {
  return api.get(
    `dashboard/${userId}/questionnaires/?limit=${limit}&offset=${offset}`
  );
};

export const getQuestionnaire = ({
  userId,
  submissionId,
}: {
  userId: number;
  submissionId: string;
}) => {
  return api.get(`/dashboard/${userId}/questionnaires/${submissionId}/`);
};

export const requestResetPassword = (username: string) => {
  return api.post('/dashboard/request-reset-password/', { username });
};

export const resetPassword = ({
  password,
  token,
}: {
  password: string;
  token: string;
}) => {
  return api.post('/dashboard/reset-password/', { password, token });
};

export const changePassword = ({
  currentPassword,
  newPassword,
}: {
  currentPassword: string;
  newPassword: string;
}) => {
  return api.post('/dashboard/change-password/', {
    current: currentPassword,
    new: newPassword,
  });
};

export default api;
