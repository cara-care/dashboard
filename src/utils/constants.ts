export const INTERCOM_APP_ID = process.env.REACT_APP_INTERCOM_APP_ID;
export const LOCALES = [
  { code: 'en', locale: 'English' },
  { code: 'de', locale: 'Deutsch' },
];
export const getHost = () => {
  return process.env.NODE_ENV === 'production'
    ? 'https://web.cara.care'
    : 'https://localhost:3000';
};
