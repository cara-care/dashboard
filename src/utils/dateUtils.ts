import { addDays, format as formatDateImpl, parseISO, format } from 'date-fns';
import { enUS, de } from 'date-fns/locale';

export const newDate = (days: number, referenceDate: Date = new Date()) => {
  const day = addDays(referenceDate, days);
  day.setHours(0, 0, 0, 0);
  return day;
};

export function getTime(date: Date): string;
export function getTime(isoString: string): string;
export function getTime(dateOrIsoString: Date | string): string;
export function getTime(dateOrIsoString: Date | string) {
  return format(getDate(dateOrIsoString), 'h:mm a');
}

export function formatDate(
  isoString: string,
  format: string,
  locale: string
): string;
export function formatDate(date: Date, format: string, locale: string): string;
export function formatDate(
  dateOrIsoString: Date | string,
  format: string,
  locale: string
) {
  return formatDateImpl(getDate(dateOrIsoString), getFormat(format), {
    locale: getLocale(locale),
  });
}

export function getLocale(localeCode: string) {
  return localeCode === 'de' ? de : enUS;
}

export function isSameOrAfter(left: Date, right: Date) {
  return left.valueOf() >= right.valueOf();
}

export function isSameOrBefore(left: Date, right: Date) {
  return left.valueOf() <= right.valueOf();
}

function getDate(dateOrIsoString: Date | string) {
  return typeof dateOrIsoString === 'string'
    ? parseISO(dateOrIsoString)
    : dateOrIsoString;
}

function getFormat(formatOrPreset: string) {
  return presets[formatOrPreset] ?? formatOrPreset;
}

export function padWith0(date: string, maxLength: number = 8) {
  const length = Math.min(date.length + 1, maxLength);
  return date.padStart(length, '0');
}

/**
 * Only include formats that are _**invalid**_ `date-fns` formats, as to not disallow usages of `getFormat` with a built-in, valid format.
 */
const presets = {
  // LLL is not a valid date-fns format, so it should be fine to use here
  LLL: 'MMMM d yyyy p',
};
