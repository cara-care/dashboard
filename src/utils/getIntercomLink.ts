const getIntercomLink = (email: string) => {
  return `https://app.intercom.io/apps/${
    process.env.REACT_APP_INTERCOM_APP_ID
  }/users/show?email=${encodeURIComponent(email)}`;
};

export default getIntercomLink;
