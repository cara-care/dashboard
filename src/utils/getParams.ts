/**
 * Extracts query string from an url
 * @example
 * getParams('https://example.com/?foo=true&baz=false')
 * // returns {foo: "true", baz: "false"}
 */
export default function getParams(url: string): { [key: string]: string } {
  const normalizedUrl = /\?/.test(url) ? url.replace(/^[^?]*./, '') : url;
  return normalizedUrl
    .split('&')
    .filter(Boolean)
    .map((el) => el.split('='))
    .reduce((acc, cur) => ({ ...acc, [cur[0]]: cur[1] }), {});
}
