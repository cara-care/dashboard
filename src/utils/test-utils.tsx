import React from 'react';
import { render } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import configureStore from './store';

export const renderWithReactIntl = (ui: React.ReactElement) => {
  return render(
    <IntlProvider locale="en" messages={require('../locale/en.json')}>
      {ui}
    </IntlProvider>
  );
};

export function renderWithRedux(
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = configureStore({
      preloadedLocale: 'en',
      preloadedTheme: undefined,
      preloadedState,
    }),
    ...renderOptions
  } = {}
) {
  function Wrapper({
    children,
  }: React.PropsWithChildren<{}>): React.ReactElement {
    return <Provider store={store}>{children}</Provider>;
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions });
}
