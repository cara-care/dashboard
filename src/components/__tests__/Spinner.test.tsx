import React from 'react';
import { renderWithReactIntl } from '../../utils/test-utils';
import Spinner from '../Spinner';

describe('<Spinner />', () => {
  it('renders correctly with default props', () => {
    const { container } = renderWithReactIntl(<Spinner />);
    expect(container).toMatchSnapshot();
  });

  it('renders correctly with custom props', () => {
    const { container } = renderWithReactIntl(<Spinner noText size={32} />);
    expect(container).toMatchSnapshot();
  });
});
