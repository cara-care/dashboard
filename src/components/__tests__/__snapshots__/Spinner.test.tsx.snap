// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Spinner /> renders correctly with custom props 1`] = `
<div>
  <div
    class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
    role="progressbar"
    style="width: 32px; height: 32px; margin-bottom: 12px;"
  >
    <svg
      class="MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
        cx="44"
        cy="44"
        fill="none"
        r="20.2"
        stroke-width="3.6"
      />
    </svg>
  </div>
</div>
`;

exports[`<Spinner /> renders correctly with default props 1`] = `
<div>
  <div
    class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
    role="progressbar"
    style="width: 40px; height: 40px; margin-bottom: 12px;"
  >
    <svg
      class="MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
        cx="44"
        cy="44"
        fill="none"
        r="20.2"
        stroke-width="3.6"
      />
    </svg>
  </div>
  <p
    class="MuiTypography-root MuiTypography-body1"
  >
    Loading data
  </p>
</div>
`;
